std/internal/cstring.d

```diff
@@ -156,178 +211,15 @@ nothrow @nogc unittest
     assert("abc".tempCString!wchar().buffPtr.asArray == "abc"w);
-}
-
-// Test for Issue 13367: ensure there is no memory corruption
-nothrow @nogc unittest
-{
-    @property str(C)() { C[300] arr = 'a'; return arr; }
-    assert(str!char.tempCString!wchar().asArray == str!wchar);
-    assert(str!char.tempCString!dchar().asArray == str!dchar);
-}
-
-version(Windows)
-    alias tempCStringW = tempCString!(wchar, char);
-
-
-private:
-
 
-// Helper UTF functions.
-// ----------------------------------------------------------------------------------------------------
-
-/**
-Returns maximum possible length of string conversion
-to another Unicode Transformation Format result.
-*/
-size_t maxLength(To, From)(in size_t length) pure nothrow @nogc
-if(isSomeChar!To && isSomeChar!From)
-{
-    static if (To.sizeof >= From.sizeof)
-        enum k = 1; // worst case: every code unit represents a character
-    else static if (To.sizeof == 1 && From.sizeof == 2)
-        enum k = 3; // worst case: every wchar in top of BMP
-    else static if (To.sizeof == 1 && From.sizeof == 4)
-        enum k = 4; // worst case: every dchar not in BMP
-    else static if (To.sizeof == 2 && From.sizeof == 4)
-        enum k = 2; // worst case: every dchar not in BMP
-    else
-        static assert(0);
-    return length * k;
-}
-
-/// ditto
-size_t maxLength(To, From)(in From[] str) pure nothrow @nogc
-{ return maxLength!(To, From)(str.length); }
-
-pure nothrow @nogc unittest
-{
-    assert(maxLength!char("abc") == 3);
-    assert(maxLength!dchar("abc") == 3);
-    assert(maxLength!char("abc"w) == 9);
-    assert(maxLength!char("abc"d) == 12);
-    assert(maxLength!wchar("abc"d) == 6);
+    import std.utf : byChar, byWchar;
+    char[300] abc = 'a';
+    assert(tempCString(abc[].byChar).buffPtr.asArray == abc);
+    assert(tempCString(abc[].byWchar).buffPtr.asArray == abc);
 }
 
 
-/**
-Copies text from $(D source) to $(D buff) performing conversion
-to different Unicode Transformation Format if needed.
-
-$(D buff) must be large enough to hold the result.
-
-Returns:
-Slice of the provided buffer $(D buff) with the copy of $(D source).
-*/
-To[] copyEncoded(To, From)(in From[] source, To[] buff) @trusted nothrow @nogc
-if(isSomeChar!To && isSomeChar!From)
-{
-    static if(is(Unqual!To == Unqual!From))
-    {
-        return buff[0 .. source.length] = source[];
-    }
-    else
-    {
-        import std.utf : byChar, byWchar, byDchar;
-        alias GenericTuple(Args...) = Args;
-        alias byFunc = GenericTuple!(byChar, byWchar, null, byDchar)[To.sizeof - 1];
-
-        To* ptr = buff.ptr;
-        const To* last = ptr + buff.length;
-        foreach(const c; byFunc(source))
-        {
-            assert(ptr != last);
-            *ptr++ = c;
-        }
-        return buff[0 .. ptr - buff.ptr];
-    }
-}
-
-///
-pure nothrow @nogc unittest
-{
-    const str = "abc-ЭЮЯ";
-    wchar[100] wsbuff;
-    assert(copyEncoded(str, wsbuff) == "abc-ЭЮЯ"w);
```

Comment:
It defeats the advantage of encapsulation to copy the unit tests of a function into every function that calls it. There isn't any code in tempCString that is peculiar to UTF. If you do not believe that byUTF is properly tested, add some more unit tests for it in the tests for byUTF. tempCString should not care about details of UTF encoding.